### PR TITLE
PLAT-1347 Option to Move Workflow Items From One Node to Another

### PR DESCRIPTION
**Pull request by thees-k**
The button for moving workflow items is displayed only if the workflow node contains items.

The user needs administration rights.

The two dropdowns in the popup are coupled: The selected workflow version determines the set of selectable nodes.

The preselected workflow version is the current version.

The `WorkflowItemsMover` is JUnit tested, the functionality of the popup was tested with WebServiceTestServer and test database. 

image

image

### CODE REVIEW

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/move/WorkflowItemsMover.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+			String itemMessageText = createItemMessageText(targetNode);
+			moveItemsToNode(targetNode, itemMessageText);
+		}
+	}
+
+	private String createItemMessageText(AGWorkflowNode targetNode) {
+
+		return String
+			.format(
+				"Item was moved from node '%s' [%s] of workflow version ID %s to node '%s' [%s] of workflow version ID %s.",
```
Comment (viktor-nikolaus) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/item/move/WorkflowItemsMover.java:37]: Maybe I18n and translate into business unit language?

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/node/items/move/WorkflowNodeMoveItemsPopup.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+	private final AGWorkflowNode sourceNode;
+	private final DomEntitySelect<AGWorkflowNode> nodeSelect;
+
+	public WorkflowNodeMoveItemsPopup(AGWorkflowNode sourceNode) {
+
+		this.sourceNode = sourceNode;
+		this.nodeSelect = new DomEntitySelect<>();
+		setCaption(WorkflowI18n.MOVE_WORKFLOW_ITEMS_TO_ANOTHER_WORKFLOW_NODE);
+		setSubCaption(WorkflowI18n.ARG1_ITEM_S_OF_SOURCE_WORKFLOW_NODE_ARG2.toDisplay(sourceNode.getAllWorkflowItems().size(), sourceNode.toDisplay()));
```
Comment (viktor-nikolaus) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/node/items/move/WorkflowNodeMoveItemsPopup.java:36]: The count could be outdated when the user clicks OK. Maybe workflow items should be loaded here and passed to the WorkflowItemsMover, so we ensure, that we only move as many items as we say here.
But some items could have been processed in the meantime as well...

Alternative:
If the user (of this action) can see how many items are currently in this node, we don't need to include the count in the subcaption.

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/node/items/move/WorkflowNodeMoveItemsPopup.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import com.softicar.platform.workflow.module.workflow.version.AGWorkflowVersion;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+class WorkflowNodeMoveItemsPopup extends DomPopup {
+
+	private final AGWorkflowNode sourceNode;
+	private final DomEntitySelect<AGWorkflowNode> nodeSelect;
```
Comment (viktor-nikolaus) [inline platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/node/items/move/WorkflowNodeMoveItemsPopup.java:29]: New screenshots? :)

File: platform-workflow-module/src/test/java/com/softicar/platform/workflow/module/workflow/item/move/WorkflowItemsMoverTest.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+		new WorkflowItemsMover(rootNode).moveItemsToNode(nodeA);
+	}
+
+	@Test(expected = SofticarUserException.class)
+	public void testMoveItemsWithSourceEqualsTarget() {
+
+		assertCount(3, AGWorkflowItem.createSelect().where(AGWorkflowItem.WORKFLOW_NODE.isEqual(rootNode)));
+
+		new WorkflowItemsMover(rootNode).moveItemsToNode(rootNode);
+	}
```
Comment (viktor-nikolaus) [inline platform-workflow-module/src/test/java/com/softicar/platform/workflow/module/workflow/item/move/WorkflowItemsMoverTest.java:125]: Maybe `assertException` or `assertExceptionMessage`?

File: platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/node/items/move/WorkflowNodeMoveItemsPopup.java (hunk 2 above)
Comment (daniel-klose) [reply platform-workflow-module/src/main/java/com/softicar/platform/workflow/module/workflow/node/items/move/WorkflowNodeMoveItemsPopup.java:36]: i dont see a big problem with having it in the subcaption, its used by admins only anyways

**Review by viktor-nikolaus — APPROVED** [review]

**Pull request merged**